File: mobile_app/lib/models/inventory_item.dart
class InventoryItemModel {
  final String id;
  final String name;
  final String? serialNumber;
  final bool hasNoSerialNumber;
  final String? brand;
  final int quantity;
  final String condition; // 'Baik' or 'Rusak'
  final String? photoPath;
  final String? categoryId;
  final String? buildingId;
  final String? roomId;
  final String? functionId;
  final String? notes;
  final String? createdBy;
  final String? createdAt;
  final String? updatedAt;

  // Relations
  final Map<String, dynamic>? category;
  final Map<String, dynamic>? building;
  final Map<String, dynamic>? room;
  final Map<String, dynamic>? itemFunction;
  final Map<String, dynamic>? creator;

  InventoryItemModel({
    required this.id,
    required this.name,
    this.serialNumber,
    required this.hasNoSerialNumber,
    this.brand,
    required this.quantity,
    required this.condition,
    this.photoPath,
    this.categoryId,
    this.buildingId,
    this.roomId,
    this.functionId,
    this.notes,
    this.createdBy,
    this.createdAt,
    this.updatedAt,
    this.category,
    this.building,
    this.room,
    this.itemFunction,
    this.creator,
  });

  bool get isGoodCondition => condition == 'Baik';
  bool isOwnedBy(String? userId) => userId != null && createdBy == userId;

  factory InventoryItemModel.fromJson(Map<String, dynamic> json) {
    return InventoryItemModel(
      id: json['id']?.toString() ?? '',
      name: json['name'] ?? '',
      serialNumber: json['serial_number'],
      hasNoSerialNumber: json['has_no_serial_number'] == true || json['has_no_serial_number'] == 1,
      brand: json['brand'],
      quantity: json['quantity'] is int ? json['quantity'] : int.tryParse(json['quantity']?.toString() ?? '1') ?? 1,
      condition: json['condition'] ?? 'Baik',
      photoPath: json['photo_path'],
      categoryId: json['category_id']?.toString(),
      buildingId: json['building_id']?.toString(),
      roomId: json['room_id']?.toString(),
      functionId: json['function_id']?.toString(),
      notes: json['notes'],
      createdBy: json['created_by']?.toString(),
      createdAt: json['created_at'],
      updatedAt: json['updated_at'],
      category: json['category'] is Map<String, dynamic> ? json['category'] : null,
      building: json['building'] is Map<String, dynamic> ? json['building'] : null,
      room: json['room'] is Map<String, dynamic> ? json['room'] : null,
      itemFunction: json['item_function'] is Map<String, dynamic> ? json['item_function'] : null,
      creator: json['creator'] is Map<String, dynamic> ? json['creator'] : null,
    );
  }
}

File: app/Models/InventoryItem.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Concerns\HasUuids;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\SoftDeletes;

class InventoryItem extends Model
{
    use HasFactory, HasUuids, SoftDeletes;

    /**
     * The "type" of the primary key ID.
     *
     * @var string
     */
    protected $keyType = 'string';

    /**
     * Indicates if the IDs are auto-incrementing.
     *
     * @var bool
     */
    public $incrementing = false;

    /**
     * The attributes that are mass assignable.
     *
     * @var array<int, string>
     */
    protected $fillable = [
        'name',
        'serial_number',
        'has_no_serial_number',
        'brand',
        'quantity',
        'photo_path',
        'condition',
        'category_id',
        'building_id',
        'room_id',
        'function_id',
        'created_by',
        'updated_by',
        'notes',
    ];

    /**
     * The attributes that should be cast.
     *
     * @var array<string, string>
     */
    protected $casts = [
        'has_no_serial_number' => 'boolean',
        'quantity' => 'integer',
        'created_at' => 'datetime',
        'updated_at' => 'datetime',
        'deleted_at' => 'datetime',
    ];

    /**
     * Kategori Barang
     */
    public function category(): BelongsTo
    {
        return $this->belongsTo(ItemCategory::class, 'category_id');
    }

    /**
     * Lokasi Gedung
     */
    public function building(): BelongsTo
    {
        return $this->belongsTo(Building::class, 'building_id');
    }

    /**
     * Lokasi Ruangan
     */
    public function room(): BelongsTo
    {
        return $this->belongsTo(Room::class, 'room_id');
    }

    /**
     * Fungsi Barang
     */
    public function itemFunction(): BelongsTo
    {
        return $this->belongsTo(ItemFunction::class, 'function_id');
    }

    /**
     * Admin yang Menginput Barang (Audit Trail)
     */
    public function creator(): BelongsTo
    {
        return $this->belongsTo(User::class, 'created_by');
    }

    /**
     * Admin yang Mengupdate Barang (Audit Trail)
     */
    public function updater(): BelongsTo
    {
        return $this->belongsTo(User::class, 'updated_by');
    }
}
